yt_dlp/extractor/crunchyroll.py

```diff
@@ -35,74 +44,177 @@ def _perform_login(self, username, password):
-            data=urllib.parse.urlencode({
+            data=urlencode_postdata({
                 'account': username,
                 'password': password,
                 'session_id': session_id
-            }).encode('ascii'))
+            }))
         if login_response['code'] != 'ok':
             raise ExtractorError('Login failed. Server message: %s' % login_response['message'], expected=True)
         if not self.is_logged_in:
             raise ExtractorError('Login succeeded but did not set etp_rt cookie')
 
-    def _get_embedded_json(self, webpage, display_id):
-        initial_state = self._parse_json(self._search_regex(
-            r'__INITIAL_STATE__\s*=\s*({.+?})\s*;', webpage, 'initial state'), display_id)
-        app_config = self._parse_json(self._search_regex(
-            r'__APP_CONFIG__\s*=\s*({.+?})\s*;', webpage, 'app config'), display_id)
-        return initial_state, app_config
-
-    def _get_params(self, lang):
-        if not CrunchyrollBaseIE.params:
-            if self._get_cookies(f'https://www.crunchyroll.com/{lang}').get('etp_rt'):
-                grant_type, key = 'etp_rt_cookie', 'accountAuthClientId'
+    def _update_query(self, lang):
+        if lang in CrunchyrollBaseIE._QUERY:
+            return
+
+        webpage = self._download_webpage(
+            f'{self._BASE_URL}/{lang}', None, note=f'Retrieving main page (lang={lang or None})')
+
+        initial_state = self._search_json(r'__INITIAL_STATE__\s*=', webpage, 'initial state', None)
+        CrunchyrollBaseIE._QUERY[lang] = traverse_obj(initial_state, {
+            'locale': ('localization', 'locale'),
+        }) or None
+
+        if CrunchyrollBaseIE._BASIC_AUTH:
+            return
+
+        app_config = self._search_json(r'__APP_CONFIG__\s*=', webpage, 'app config', None)
+        cx_api_param = app_config['cxApiParams']['accountAuthClientId' if self.is_logged_in else 'anonClientId']
+        self.write_debug(f'Using cxApiParam={cx_api_param}')
+        CrunchyrollBaseIE._BASIC_AUTH = 'Basic ' + base64.b64encode(f'{cx_api_param}:'.encode()).decode()
+
+    def _update_auth(self):
+        if CrunchyrollBaseIE._AUTH_HEADERS and CrunchyrollBaseIE._AUTH_REFRESH > time_seconds():
+            return
+
+        assert CrunchyrollBaseIE._BASIC_AUTH, '_update_query needs to be called at least one time beforehand'
+        grant_type = 'etp_rt_cookie' if self.is_logged_in else 'client_id'
+        auth_response = self._download_json(
+            f'{self._BASE_URL}/auth/v1/token', None, note=f'Authenticating with grant_type={grant_type}',
+            headers={'Authorization': CrunchyrollBaseIE._BASIC_AUTH}, data=f'grant_type={grant_type}'.encode())
+
+        CrunchyrollBaseIE._AUTH_HEADERS = {'Authorization': auth_response['token_type'] + ' ' + auth_response['access_token']}
+        CrunchyrollBaseIE._AUTH_REFRESH = time_seconds(seconds=traverse_obj(auth_response, ('expires_in', {float_or_none}), default=300) - 10)
```

Comment:
Not mandatory, but I prefer using self where applicable.

```suggestion
        if lang in self._QUERY:
            return

        webpage = self._download_webpage(
            f'{self._BASE_URL}/{lang}', None, note=f'Retrieving main page (lang={lang or None})')

        initial_state = self._search_json(r'__INITIAL_STATE__\s*=', webpage, 'initial state', None)
        self._QUERY[lang] = traverse_obj(initial_state, {
            'locale': ('localization', 'locale'),
        }) or None

        if self._BASIC_AUTH:
            return

        app_config = self._search_json(r'__APP_CONFIG__\s*=', webpage, 'app config', None)
        cx_api_param = app_config['cxApiParams']['accountAuthClientId' if self.is_logged_in else 'anonClientId']
        self.write_debug(f'Using cxApiParam={cx_api_param}')
        self._BASIC_AUTH = 'Basic ' + base64.b64encode(f'{cx_api_param}:'.encode()).decode()

    def _update_auth(self):
        if self._AUTH_HEADERS and self._AUTH_REFRESH > time_seconds():
            return

        assert self._BASIC_AUTH, '_update_query needs to be called at least one time beforehand'
        grant_type = 'etp_rt_cookie' if self.is_logged_in else 'client_id'
        auth_response = self._download_json(
            f'{self._BASE_URL}/auth/v1/token', None, note=f'Authenticating with grant_type={grant_type}',
            headers={'Authorization': self._BASIC_AUTH}, data=f'grant_type={grant_type}'.encode())

        self._AUTH_HEADERS['Authorization'] = f'{auth_response["token_type"]} {auth_response["access_token"]}'
        CrunchyrollBaseIE._AUTH_REFRESH = time_seconds(seconds=traverse_obj(auth_response, ('expires_in', {float_or_none}), default=300) - 10)
```

Reply:
I prefer to use the class name if we are doing class operations, so its more clear as to where we are writing that to/reading that from. `self` doesn't tell us much, could be anywhere in MRO